File: rails/app/models/card_price_history.rb
class CardPriceHistory < ApplicationRecord
  self.table_name = 'card_price_histories'

  belongs_to :card, class_name: 'Card'

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:price_bounds_consistent, 'min_price <= avg_price <= max_price must hold') unless (((min_price.nil? || (!avg_price.nil? && min_price.to_f <= avg_price.to_f)) && (avg_price.nil? || (!max_price.nil? && avg_price.to_f <= max_price.to_f))))
  end

  def to_s
    price_date.to_s
  end

  # Business operations

  def price_change_percent(previous_avg)
    raise NotImplementedError, "price_change_percent not implemented"
  end

  def is_price_spike(threshold_percent)
    raise NotImplementedError, "is_price_spike not implemented"
  end
end

File: laravel/app/Models/Marketplace/CardPriceHistory.php
<?php

namespace App\Models\Marketplace;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use App\Models\Cards\Card;

class CardPriceHistory extends Model
{
    protected $table = 'card_price_histories';

    protected $fillable = ['price_date', 'avg_price', 'min_price', 'max_price', 'volume', 'foil', 'card_id'];

    protected $casts = [
        'price_date' => 'date',
        'avg_price' => 'decimal:2',
        'min_price' => 'decimal:2',
        'max_price' => 'decimal:2',
        'foil' => 'boolean',
    ];

    public function card(): BelongsTo
    {
        return $this->belongsTo(Card::class, 'card_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!((($this->min_price === null || ($this->avg_price !== null && (float)$this->min_price <= (float)$this->avg_price)) && ($this->avg_price === null || ($this->max_price !== null && (float)$this->avg_price <= (float)$this->max_price))))) {
            $errors['price_bounds_consistent'] = 'min_price <= avg_price <= max_price must hold';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function priceChangePercent($previous_avg): string
    {
        throw new \RuntimeException('price_change_percent not implemented');
    }

    public function isPriceSpike($threshold_percent): bool
    {
        throw new \RuntimeException('is_price_spike not implemented');
    }

}
